File: src/Java/myservicebus-abstractions/src/main/java/com/myservicebus/EntityNameFormatter.java
package com.myservicebus;

import java.lang.reflect.Proxy;

public final class EntityNameFormatter {
    private static MessageEntityNameFormatter formatter = new DefaultMessageEntityNameFormatter();

    private EntityNameFormatter() { }

    public static MessageEntityNameFormatter getFormatter() {
        return formatter;
    }

    public static void setFormatter(MessageEntityNameFormatter f) {
        formatter = f;
    }

    public static String format(Class<?> messageType) {
        EntityName attr = messageType.getAnnotation(EntityName.class);
        if (attr != null)
            return attr.value();
        return formatter.formatEntityName(messageType);
    }

    static class DefaultMessageEntityNameFormatter implements MessageEntityNameFormatter {
        @Override
        public String formatEntityName(Class<?> messageType) {
            if (Proxy.isProxyClass(messageType) && messageType.getInterfaces().length > 0) {
                messageType = messageType.getInterfaces()[0];
            }
            return String.format("TestApp:%s", messageType.getSimpleName());
        }
    }
}

File: src/MyServiceBus/EntityNameFormatter.cs
using System;
using System.Reflection;

namespace MyServiceBus;

public static class EntityNameFormatter
{
    static IMessageEntityNameFormatter _formatter = new DefaultMessageEntityNameFormatter();

    public static IMessageEntityNameFormatter Formatter => _formatter;

    public static void SetFormatter(IMessageEntityNameFormatter formatter)
    {
        _formatter = formatter;
    }

    [Throws(typeof(AmbiguousMatchException), typeof(TypeLoadException))]
    public static string Format(Type messageType)
    {
        var attr = messageType.GetCustomAttribute<EntityNameAttribute>();
        if (attr != null)
            return attr.EntityName;

        return _formatter.FormatEntityName(messageType);
    }

    class DefaultMessageEntityNameFormatter : IMessageEntityNameFormatter
    {
        public string FormatEntityName(Type messageType)
        {
            return $"{messageType.Namespace}:{messageType.Name}";
        }
    }
}
